conans/client/loader.py

```diff
@@ -43,17 +43,23 @@ def invalidate_caches(self):
         self.cached_conanfiles = {}
         self._python_requires.invalidate_caches()
 
-    def load_class(self, conanfile_path):
-        conanfile = self.cached_conanfiles.get(conanfile_path)
-        if conanfile:
-            return conanfile
+    def load_class(self, conanfile_path, lock_python_requires=None):
+        cached = self.cached_conanfiles.get(conanfile_path)
+        if cached and cached[1] == lock_python_requires:
+            return cached[0]
 
+        if lock_python_requires is not None:
+            self._python_requires.locked_versions = {r.name: r for r in lock_python_requires}
         try:
             self._python_requires.valid = True
             _, conanfile = parse_conanfile(conanfile_path, self._python_requires)
             self._python_requires.valid = False
-            self.cached_conanfiles[conanfile_path] = conanfile
+
+            self._python_requires.locked_versions = None
+            self.cached_conanfiles[conanfile_path] = (conanfile, lock_python_requires)
```

Comment:
The `conanfile` object already stores the python requires list/dict, `conanfile.python_requires`. I know it makes a little bit more difficult the above getter, but we won't duplicate info.